pydomcfg/domzgr/zco.py

```diff
@@ -0,0 +1,224 @@
+                (ln_e3_dep=False); for backward compatibility with v3.6.
+             2) discrete derivative (central-difference) of levels' depth
+                (ln_e3_dep=True). The only possibility from v4.0.
+
+    References:
+       *) NEMO v4.0 domzgr/zgr_z subroutine
+       *) Marti, Madec & Delecluse, 1992, JGR, 97, No8, 12,763-12,766.
+    """
+
+    # --------------------------------------------------------------------------
+    def generate(
+        self,
+        ppdzmin,
+        pphmax,
+        ppkth: float = 0.0,
+        ppacr: int = 0,
+        ppsur: float = pp_to_be_computed,
+        ppa0: float = pp_to_be_computed,
+        ppa1: float = pp_to_be_computed,
+        ldbletanh: bool = False,
+        ppa2: float = 0.0,
+        ppkth2: float = 0.0,
+        ppacr2: float = 0.0,
+    ) -> Dataset:
+        """
+        Generate NEMO geopotential z-coordinates model levels.
+
+        Parameters
+        ----------
+        ppdzmin: float
+            Minimum thickness for the top layer (units: m)
+        pphmax: float
+            Depth of the last W-level (total depth of the ocean basin) (units: m)
+        ppacr: int
+            Stretching factor, nondimensional > 0. The larger ppacr,
+            the smaller the stretching. Values from 3 to 10 are usual.
+            (default = 0., no stretching, uniform grid)
+        ppkth: float
+            Model level at which approximately max stretching occurs.
+            Nondimensional > 0, usually of order 1/2 or 2/3 of jpk.
+            (default = 0., i.e. no stretching, uniform grid)
+        ppsur: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ppa0: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ppa1: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ldbletanh: bool
+            Logical flag to use or not double tanh stretching function (default = False)
+        ppa2: float
+            Double tanh stretching function parameter
+        ppkth2: float
+            Double tanh stretching function parameter
+        ppacr2: float
+            Double tanh stretching function parameter
+
+        Returns
+        -------
+        Dataset
+           Describing the 3D geometry of the model
+
+        """
+        self._ppdzmin = ppdzmin
+        self._pphmax = pphmax
+        self._ppkth = ppkth
+        self._ppacr = ppacr
+        self._ppsur = ppsur
+        self._ppa0 = ppa0
+        self._ppa1 = ppa1
+        self._ldbletanh = ldbletanh
+        self._ppa2 = ppa2
+        self._ppkth2 = ppkth2
+        self._ppacr2 = ppacr2
```

Comment:
I would place these in init, the same you did for `Zgr`.
Init can look like this:
```python
class Zco(Zgr):
    def __init__(self,  ds_bathy: Dataset, jpk: int, ppdzmin, pphmax, ....):
        self._ppdzmin = ppdzmin
        self._pphmax = pphmax
        ....
        # Initialize base class
        super().__init__(ds_bathy, jpk)  # This will fill self._ds_bathy and self._jpk
```

Reply:
Yes, I considered this option but I think it is cleaner as it is. 

Let me explain:

from pydomcfg.tests.bathymetry import Bathymetry
from pydomcfg.domzgr.zco import Zco

# Create the hor grid and bathy dataset
ds = Bathymetry(1000.,1200.,100,200).flat(5000.)

# Create the z-coordinates grid generator: at this point we 
# have to specify only the number of levels we want to use
zco = Zco(ds,31)

# Generate the grid
dsz = zco.generate(10., 5000., 21.43336197938, 3, -4762.96143546300, 255.58049070440, 245.58132232490, False)

In this way, with one grid generator we can create how many configurations we want of a vertical grid with jpk levels. In the other way you have to recreate the grid generator every time you'll change the parameters.

Reply:
I see, makes sense! 
All methods that use those parameters should probably become private because they would fail if executed before generate (e.g., `compute_pp` -> `_compute_pp`). 
Do you think `generate` is going to be the only public method? Then I guess rather than `generate`, we could probably replace `generate` with `__call__`, and the syntax becomes:
```python
ds = Bathymetry(1000., 1200., 100, 200).flat(5000.)
zco = Zco(ds, 31)
dsz = zco(10., 5000., 21.43336197938, 3, -4762.96143546300, 255.58049070440, 245.58132232490, False)
```

This means that `__call__` will become an abstract method in the parent class, and the only constraint in the child class is that they define `__call__`.

Reply:
1. Yes I agree about making those methods private
2. I really like the idea of using the __call__ method. I am not sure if generate will be the only public method - for now I would say yes, but I don't know where we will end up :) ... but I guess we will always able to come back, so yes, let's do it! :)